src/SourceTerms/LightBulb/CPU_Src_Lightbulb.cpp

```diff
@@ -0,0 +1,403 @@
+                           const EoS_t *EoS, const double AuxArray_Flt[], const int AuxArray_Int[] )
+{
+
+// check
+#  ifdef GAMER_DEBUG
+   if ( AuxArray_Flt == NULL )   printf( "ERROR : AuxArray_Flt == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Int == NULL )   printf( "ERROR : AuxArray_Int == NULL in %s !!\n", __FUNCTION__ );
+#  endif
+
+
+   const real Kelvin2MeV = AuxArray_Flt[SRC_AUX_KELVIN2MEV];
+   const real sEint2Code = AuxArray_Flt[SRC_AUX_VSQR2CODE ];
+
+   const real x0         = x - SrcTerms->BoxCenter[0];
+   const real y0         = y - SrcTerms->BoxCenter[1];
+   const real z0         = z - SrcTerms->BoxCenter[2];
+   const real r2_CGS     = SQR( SrcTerms->Unit_L ) * (  SQR( x0 ) + SQR( y0 ) + SQR( z0 )  );
+
+#  ifdef MHD
+   const real Emag       = (real)0.5*(  SQR( B[MAGX] ) + SQR( B[MAGY] ) + SQR( B[MAGZ] )  );
+#  else
+   const real Emag       = NULL_REAL;
+#  endif
+
+   const real Dens_Code  = fluid[DENS];
+   const real Eint_Code  = Hydro_Con2Eint( fluid[DENS], fluid[MOMX], fluid[MOMY], fluid[MOMZ], fluid[ENGY],
+                                           true, MinEint, Emag );
+   const real Ye         = fluid[YE] / fluid[DENS];
+
+
+// 1. call nuclear EoS driver to obtain Xn, Xp, and temperature, using energy mode
+#  if ( NUC_TABLE_MODE == NUC_TABLE_MODE_TEMP )
+   const int  NTarget = 2;
+#  else
+   const int  NTarget = 3;
+#  endif
+         int  In_Int[NTarget+1];
+         real In_Flt[3], Out[NTarget+1];
+
+   In_Flt[0] = Dens_Code;
+   In_Flt[1] = Eint_Code;
+   In_Flt[2] = Ye;
+
+   In_Int[0] = NTarget;
+   In_Int[1] = NUC_VAR_IDX_XN;
+   In_Int[2] = NUC_VAR_IDX_XP;
+#  if ( NUC_TABLE_MODE == NUC_TABLE_MODE_ENGY )
+   In_Int[3] = NUC_VAR_IDX_EORT;
+#  endif
+
+
+   EoS->General_FuncPtr( NUC_MODE_ENGY, Out, In_Flt, In_Int, EoS->AuxArrayDevPtr_Flt, EoS->AuxArrayDevPtr_Int, EoS->Table );
+
+   const real Xn       = Out[0];               // neutron number fraction
+   const real Xp       = Out[1];               // proton  number fraction
+   const real Temp_MeV = Out[2] * Kelvin2MeV;  // temperature in MeV
+
+
+// 2. calculate cooling and heating rate
+   const real rate_heating = 1.544e20 * ( SrcTerms->Lightbulb_Lnue / 1.0e52 ) * ( 1.0e14 / r2_CGS )
+                           * SQR( 0.25 * SrcTerms->Lightbulb_Tnue );
+   const real rate_cooling = 1.399e20 * CUBE(  SQR( 0.5 * Temp_MeV )  );
+
+// approximate the optical depth by density in unit of 10^11 g/cm^3
+   const real tau      = 1.0e-11 * Dens_Code * SrcTerms->Unit_D;
+   const real rate_CGS = ( rate_heating - rate_cooling ) * ( Xn + Xp ) * EXP( -tau );
+
+
+// 3. calculate the change in internal energy and update the input energy density
+   const real dEint_Code  = ( rate_CGS * dt * SrcTerms->Unit_T ) * sEint2Code * Dens_Code;
+   const real Eint_Update = Eint_Code + dEint_Code;
+
+   fluid[ENGY] = Hydro_ConEint2Etot( fluid[DENS], fluid[MOMX], fluid[MOMY], fluid[MOMZ], Eint_Update, Emag );
```

Comment:
I am not sure how `Hhydro_ConEint2Etot` work. Just want make sure after the `Eint_Update` is updated, we do call the eos driver (with energy mode) to update entropy or other quantities?

Reply:
The `Hydro_ConEint2Etot()` adds the kinematic energy to internal energy to obtain the total energy. I do not update other quantities since entropy is not stored. Do you think whether we need to update the Ye using the new energy?

Reply:
then it should be fine. Ye won't change after core bounce with the lightbulb scheme.